packages/back-end/src/routers/environment/environment.controller.ts

```diff
@@ -41,13 +119,118 @@ export const putEnvironments = async (
     return addEnvironmentToOrganizationEnvironments(environment, acc, false);
   }, getEnvironments(org));
 
-  await updateOrganization(org.id, {
-    settings: {
-      ...org.settings,
-      environments: updatedEnvironments,
+  try {
+    await updateOrganization(org.id, {
+      settings: {
+        ...org.settings,
+        environments: updatedEnvironments,
+      },
+    });
+
+    await req.audit({
+      event: "environments.update",
+      entity: {
+        object: "organization",
+        id: org.id,
+      },
+      details: auditDetailsUpdate(existingEnvs, updatedEnvironments),
+    });
+
+    res.json({ environments, status: 200 });
+  } catch (e) {
+    res.status(400).json({
+      status: 400,
+      message: e.message || "An error occurred",
+    });
+  }
+};
+
+export const putEnvironment = async (
+  req: AuthRequest<
+    {
+      environment: Environment;
     },
-  });
-  res.json({ environments });
+    { id: string }
+  >,
+  res: Response
+) => {
+  const { environment } = req.body;
+  const context = getContextFromReq(req);
+  const { org } = context;
+
+  const envsArr = org.settings?.environments || [];
+
+  const existingEnvIndex = envsArr.findIndex(
+    (env) => env.id === environment.id
+  );
+
+  if (existingEnvIndex < 0) {
+    return res.status(400).json({
+      status: 400,
+      message: `Could not find environment: ${environment.id}`,
+    });
+  }
+
+  //MKTODO: Update this to canUpdate when I break canCreateOrUpdateEnvironment out into two methods
+  if (!context.permissions.canCreateOrUpdateEnvironment(environment)) {
+    context.permissions.throwPermissionError();
+  }
+
+  const updatedEnvs = [...envsArr];
+  updatedEnvs[existingEnvIndex] = environment;
+
+  try {
+    await updateOrganization(org.id, {
+      settings: {
+        ...org.settings,
+        environments: updatedEnvs,
+      },
+    });
+
+    if ("projects" in environment) {
+      const existingProjects = envsArr[existingEnvIndex].projects || [];
+      const newProjects = environment.projects || [];
```

Comment:
nit: 
```suggestion
    const  newProjects = environment.projects;
    if (newProjects) {
      const existingProjects = envsArr[existingEnvIndex].projects || [];
```
The typechecker should then know that inside this conditional, `newProjects` is actually defined. Also I remember some noise around `key in object` with old school JS but I'm not sure what the status of this is nowadays.. 😆